File: src/resilience/reaper.ts
import { syncCli } from '../workers/spawn-utils.js';

/**
 * Best-effort stale process reaper.
 * - killStaleProcessTree(pid): SIGTERM then SIGKILL the process group.
 * - findStaleWorkerPids: scan `ps` for opencode/claude/omp workers older than the threshold.
 * Guarded: only kills PIDs whose cmdline matches known worker patterns.
 */

const WORKER_PATTERN = /\b(opencode|claude|omp)(\s|$|--)/i;

function cmdlineFor(pid: number): string {
  try {
    const out = syncCli('ps', ['-o', `command=`, '-p', String(pid)], {
      cwd: '/', timeoutMs: 2_000,
    });
    return out.trim();
  } catch {
    return '';
  }
}

/**
 * PIDs of this process plus all its ancestors. Reaping must never kill
 * itself, its parent shell, or the devagent/LaunchAgent chain that invoked
 * it — otherwise a mid-run reap terminates the very pipeline it protects.
 */
export function ownAncestryPids(): Set<number> {
  const pids = new Set<number>();
  let pid = process.pid;
  for (let i = 0; i < 64 && pid > 1 && !pids.has(pid); i++) {
    pids.add(pid);
    try {
      const out = syncCli('ps', ['-o', `ppid=`, '-p', String(pid)], {
        cwd: '/', timeoutMs: 2_000,
      });
      const next = Number(out.trim());
      if (!Number.isFinite(next) || next <= 1) break;
      pid = next;
    } catch {
      break;
    }
  }
  return pids;
}

function isWorkerPid(pid: number): boolean {
  const cmd = cmdlineFor(pid);
  return WORKER_PATTERN.test(cmd);
}

export function killStaleProcessTree(pid: number, reason = 'stale'): boolean {
  if (!Number.isFinite(pid) || pid <= 1) return false;
  if (!isWorkerPid(pid)) return false;
  try {
    // Try process group first (negative pid), then direct.
    try {
      process.kill(-pid, 'SIGTERM');
    } catch {
      process.kill(pid, 'SIGTERM');
    }
    // Brief grace then SIGKILL
    const deadline = Date.now() + 800;
    while (Date.now() < deadline) {
      try {
        process.kill(pid, 0);
      } catch {
        return true; // gone
      }
      Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, 50);
    }
    try {
      process.kill(-pid, 'SIGKILL');
    } catch {
      try {
        process.kill(pid, 'SIGKILL');
      } catch {
        /* gone */
      }
    }
    return true;
  } catch {
    return false;
  }
}

export interface StaleWorker {
  pid: number;
  elapsedMs: number;
  command: string;
}

/**
 * Parse ps `etime` output ([[dd-]hh:]mm:ss or seconds) to milliseconds.
 * Portable across Linux procps and macOS BSD ps; `etimes` (raw seconds
 * column) is Linux-only and silently breaks the scan on macOS.
 */
export function parseEtimeToMs(etime: string): number {
  const s = etime.trim();
  if (/^\d+$/.test(s)) return Number(s) * 1000;
  const m = s.match(/^(?:(\d+)-)?(?:(\d+):)?(\d+):(\d+)$/);
  if (!m) return 0;
  const days = Number(m[1] ?? 0);
  const hours = Number(m[2] ?? 0);
  const minutes = Number(m[3]!);
  const seconds = Number(m[4]!);
  return (((days * 24 + hours) * 60 + minutes) * 60 + seconds) * 1000;
}

/**
 * A process is only reap-eligible when it is provably a devagent-spawned
 * worker: headless print-mode with JSON output. Interactive sessions
 * (`claude` TUI in any project, or the user's bare `omp` TUIs) never match,
 * so the reaper can never kill the user's live work (2026-08-26 incident:
 * pattern-only matching reaped unrelated interactive claude processes
 * machine-wide).
 */
const DEVAGENT_WORKER_CMD = /(^|\s)(--print|-p)(\s|$)/;
const DEVAGENT_RESUME_CMD = /(^|\s)(--continue|-c)(\s|$)/;
const OMP_MODE_JSON_CMD = /(^|\s)--mode(\s|=)json(\s|$)/;

export function isDevagentWorkerCmd(cmd: string): boolean {
  if (!WORKER_PATTERN.test(cmd)) return false;
  const headless = DEVAGENT_WORKER_CMD.test(cmd) || DEVAGENT_RESUME_CMD.test(cmd);
  if (!headless) return false;
  // claude-code / opencode: --output-format json (unchanged)
  if (/\b(opencode|claude)\b/i.test(cmd)) return /--output-format\b/.test(cmd);
  // omp: --mode json (adapter always emits `--mode json`; interactive omp has
  // no flag and never matches)
  if (/\bomp\b/.test(cmd)) return OMP_MODE_JSON_CMD.test(cmd);
  return false;
}

function cwdFor(pid: number): string {
  try {
    const out = syncCli('lsof', ['-a', '-p', String(pid), '-d', 'cwd', '-Fn'], {
      cwd: '/', timeoutMs: 2_000,
    });
    const line = out.split('\n').find((l) => l.startsWith('n'));
    return line ? line.slice(1) : '';
  } catch {
    return '';
  }
}

export function findStaleWorkerPids(
  olderThanMs = 10 * 60_000,
  opts: { cwdPrefix?: string } = {},
): StaleWorker[] {
  try {
    // etime (not etimes): BSD ps on macOS rejects etimes, which made the
    // whole scan throw and silently report "no stale workers" forever.
    const out = syncCli('ps', ['-eo', 'pid,etime,command'], {
      cwd: '/', timeoutMs: 3_000,
    });
    const lines = out.split('\n').slice(1);
    const own = ownAncestryPids();
    const stale: StaleWorker[] = [];
    for (const line of lines) {
      const m = line.trim().match(/^(\d+)\s+(\S+)\s+(.*)$/);
      if (!m) continue;
      const pid = Number(m[1]);
      if (own.has(pid)) continue;
      const cmd = m[3] ?? '';
      if (!isDevagentWorkerCmd(cmd)) continue;
      if (opts.cwdPrefix && !cwdFor(pid).startsWith(opts.cwdPrefix)) continue;
      const elapsedMs = parseEtimeToMs(m[2] ?? '');
      if (elapsedMs >= olderThanMs) stale.push({ pid, elapsedMs, command: cmd.slice(0, 300) });
    }
    return stale;
  } catch {
    return [];
  }
}

/** Kill all stale workers older than threshold. Returns pids killed. */
export function reapStaleWorkers(
  olderThanMs = 10 * 60_000,
  dryRun = false,
  opts: { cwdPrefix?: string } = {},
): StaleWorker[] {
  const stale = findStaleWorkerPids(olderThanMs, opts);
  if (dryRun) return stale;
  for (const s of stale) {
    killStaleProcessTree(s.pid, 'reap-stale');
  }
  return stale;
}

File: internal/pipeline/reaper.go
// Stale-worker reaper: Go port of src/resilience/reaper.ts.
//
// Best-effort stale process reaper. Scans `ps` for headless devagent
// worker processes older than a threshold and kills their whole process
// group. Kill decisions are always gated on the devagent headless command
// shape, so interactive operator sessions are never reap-eligible (the
// 2026-08-26 incident class).

package pipeline

import (
	"os"
	"regexp"
	"strconv"
	"strings"

	"github.com/FreePeak/devagent/internal/spawn"
)

// workerPattern mirrors the TS WORKER_PATTERN.
var workerPattern = regexp.MustCompile(`(?i)\b(opencode|claude|omp|pi)(\s|$|--)`)

// ReapOptions mirrors the TS { cwdPrefix?: string } opts object.
type ReapOptions struct {
	// CWDPrefix restricts the scan to worker processes whose cwd starts
	// with this path (the repo's .devagent-worktrees). "" = no filter.
	CWDPrefix string
}

// StaleWorker mirrors the TS StaleWorker interface.
type StaleWorker struct {
	Pid       int    `json:"pid"`
	ElapsedMs int64  `json:"elapsedMs"`
	Command   string `json:"command"`
}

// reapProbes is the test seam for the probe funcs: tests inject fake
// ps/lsof output through these; nil fields fall back to the real
// spawn.RunCli probe.
var reapProbes = struct {
	// cmdline runs `ps -o command= -p <pid>` (2s timeout).
	cmdline func(pid int) string
	// ppid runs `ps -o ppid= -p <pid>` (2s timeout); false maps the TS
	// throw.
	ppid func(pid int) (int, bool)
	// cwd runs the lsof cwd probe for pid (2s timeout).
	cwd func(pid int) string
	// scan runs `ps -eo pid,etime,command` (3s timeout); false maps the TS
	// throw (caller returns no stale workers).
	scan func() (out string, ok bool)
}{}

// reapKillTree is the test seam for the kill step (tests record instead of
// signaling real processes).
var reapKillTree = killStaleProcessTree

func probeCmdline(pid int) string {
	if reapProbes.cmdline != nil {
		return reapProbes.cmdline(pid)
	}
	out := spawn.RunCli("ps", []string{"-o", "command=", "-p", strconv.Itoa(pid)},
		spawn.Options{Dir: "/", TimeoutMs: 2_000})
	return strings.TrimSpace(out.Stdout)
}

func probePpid(pid int) (int, bool) {
	if reapProbes.ppid != nil {
		return reapProbes.ppid(pid)
	}
	out := spawn.RunCli("ps", []string{"-o", "ppid=", "-p", strconv.Itoa(pid)},
		spawn.Options{Dir: "/", TimeoutMs: 2_000})
	n, err := strconv.Atoi(strings.TrimSpace(out.Stdout))
	if err != nil {
		return 0, false
	}
	return n, true
}

// OwnAncestryPids mirrors ownAncestryPids: PIDs of this process plus all
// its ancestors. Reaping must never kill itself, its parent shell, or the
// devagent/LaunchAgent chain that invoked it — otherwise a mid-run reap
// terminates the very pipeline it protects.
func OwnAncestryPids() map[int]bool {
	pids := map[int]bool{}
	pid := os.Getpid()
	for i := 0; i < 64 && pid > 1 && !pids[pid]; i++ {
		pids[pid] = true
		next, ok := probePpid(pid)
		if !ok || next <= 1 {
			break
		}
		pid = next
	}
	return pids
}

// isWorkerPid mirrors the TS isWorkerPid.
func isWorkerPid(pid int) bool {
	return workerPattern.MatchString(probeCmdline(pid))
}

// KillStaleProcessTree is the pinned export: kill a stale worker's whole
// process tree. Returns whether the pid was deemed a killable worker.
func KillStaleProcessTree(pid int) bool {
	return killStaleProcessTree(pid)
}

var digitsOnly = regexp.MustCompile(`^\d+$`)
var etimePattern = regexp.MustCompile(`^(?:(\d+)-)?(?:(\d+):)?(\d+):(\d+)$`)

// ParseEtimeToMs mirrors parseEtimeToMs: parse ps `etime` output
// ([[dd-]hh:]mm:ss or seconds) to milliseconds. Portable across Linux
// procps and macOS BSD ps; `etimes` (raw seconds column) is Linux-only and
// silently breaks the scan on macOS.
func ParseEtimeToMs(etime string) int64 {
	s := strings.TrimSpace(etime)
	if digitsOnly.MatchString(s) {
		n, err := strconv.ParseInt(s, 10, 64)
		if err != nil {
			return 0
		}
		return n * 1000
	}
	m := etimePattern.FindStringSubmatch(s)
	if m == nil {
		return 0
	}
	days := int64(0)
	if m[1] != "" {
		days, _ = strconv.ParseInt(m[1], 10, 64)
	}
	hours := int64(0)
	if m[2] != "" {
		hours, _ = strconv.ParseInt(m[2], 10, 64)
	}
	minutes, _ := strconv.ParseInt(m[3], 10, 64)
	seconds, _ := strconv.ParseInt(m[4], 10, 64)
	return (((days*24+hours)*60+minutes)*60 + seconds) * 1000
}

// Reap-eligibility patterns mirror the TS consts: a process is only
// reap-eligible when it is provably a devagent-spawned worker: headless
// print-mode with JSON output. Interactive sessions (`claude` TUI in any
// project, or the user's bare `omp` TUIs) never match, so the reaper can
// never kill the user's live work (2026-08-26 incident: pattern-only
// matching reaped unrelated interactive claude processes machine-wide).
var (
	devagentWorkerCmd = regexp.MustCompile(`(^|\s)(--print|-p)(\s|$)`)
	devagentResumeCmd = regexp.MustCompile(`(^|\s)(--continue|-c)(\s|$)`)
	ompModeJSONCmd    = regexp.MustCompile(`(^|\s)--mode(\s|=)json(\s|$)`)
	piCmdPattern      = regexp.MustCompile(`(^|\b)pi(\s|$)`)
	opencodeClaudeRe  = regexp.MustCompile(`(?i)\b(opencode|claude)\b`)
	ompWordRe         = regexp.MustCompile(`\bomp\b`)
	outputFormatRe    = regexp.MustCompile(`--output-format\b`)
)

// IsDevagentWorkerCmd mirrors isDevagentWorkerCmd.
func IsDevagentWorkerCmd(cmd string) bool {
	if !workerPattern.MatchString(cmd) {
		return false
	}
	headless := devagentWorkerCmd.MatchString(cmd) || devagentResumeCmd.MatchString(cmd)
	if !headless {
		return false
	}
	// claude-code / opencode: --output-format json (unchanged)
	if opencodeClaudeRe.MatchString(cmd) {
		return outputFormatRe.MatchString(cmd)
	}
	// omp / pi: --mode json (adapters always emit `--mode json`; interactive
	// omp/pi have no flag and never match)
	if ompWordRe.MatchString(cmd) {
		return ompModeJSONCmd.MatchString(cmd)
	}
	if piCmdPattern.MatchString(cmd) {
		return ompModeJSONCmd.MatchString(cmd)
	}
	return false
}

// probeCwd mirrors the TS cwdFor: the lsof cwd probe (`-Fn` prints the
// path on the line starting with `n`).
func probeCwd(pid int) string {
	if reapProbes.cwd != nil {
		return reapProbes.cwd(pid)
	}
	out := spawn.RunCli("lsof", []string{"-a", "-p", strconv.Itoa(pid), "-d", "cwd", "-Fn"},
		spawn.Options{Dir: "/", TimeoutMs: 2_000})
	for _, l := range strings.Split(out.Stdout, "\n") {
		if strings.HasPrefix(l, "n") {
			return l[1:]
		}
	}
	return ""
}

// psLinePattern mirrors the TS /^(\d+)\s+(\S+)\s+(.*)$/ row matcher.
var psLinePattern = regexp.MustCompile(`^(\d+)\s+(\S+)\s+(.*)$`)

// defaultOlderThanMs mirrors the TS default threshold.
const defaultOlderThanMs = 10 * 60_000

// FindStaleWorkerPids mirrors findStaleWorkerPids. etime (not etimes): BSD
// ps on macOS rejects etimes, which made the whole scan throw and silently
// report "no stale workers" forever.
func FindStaleWorkerPids(olderThanMs int, opts *ReapOptions) []StaleWorker {
	if olderThanMs <= 0 {
		olderThanMs = defaultOlderThanMs
	}
	cwdPrefix := ""
	if opts != nil {
		cwdPrefix = opts.CWDPrefix
	}
	var raw string
	var ok bool
	if reapProbes.scan != nil {
		raw, ok = reapProbes.scan()
	} else {
		out := spawn.RunCli("ps", []string{"-eo", "pid,etime,command"},
			spawn.Options{Dir: "/", TimeoutMs: 3_000})
		raw, ok = out.Stdout, out.ExitCode == 0
	}
	if !ok {
		return []StaleWorker{}
	}
	lines := strings.Split(raw, "\n")
	if len(lines) > 0 {
		lines = lines[1:] // drop the header row like the TS slice(1)
	}
	own := OwnAncestryPids()
	stale := []StaleWorker{}
	for _, line := range lines {
		m := psLinePattern.FindStringSubmatch(strings.TrimSpace(line))
		if m == nil {
			continue
		}
		pid, err := strconv.Atoi(m[1])
		if err != nil || own[pid] {
			continue
		}
		cmd := m[3]
		if !IsDevagentWorkerCmd(cmd) {
			continue
		}
		if cwdPrefix != "" && !strings.HasPrefix(probeCwd(pid), cwdPrefix) {
			continue
		}
		elapsedMs := ParseEtimeToMs(m[2])
		if elapsedMs >= int64(olderThanMs) {
			command := cmd
			if len(command) > 300 {
				command = command[:300]
			}
			stale = append(stale, StaleWorker{Pid: pid, ElapsedMs: elapsedMs, Command: command})
		}
	}
	return stale
}

// ReapStaleWorkers mirrors reapStaleWorkers: kill all stale workers older
// than the threshold. Returns the stale set (signaled when !dryRun).
func ReapStaleWorkers(olderThanMs int, dryRun bool, opts *ReapOptions) []StaleWorker {
	if olderThanMs <= 0 {
		olderThanMs = defaultOlderThanMs
	}
	stale := FindStaleWorkerPids(olderThanMs, opts)
	if dryRun {
		return stale
	}
	for _, s := range stale {
		reapKillTree(s.Pid)
	}
	return stale
}
